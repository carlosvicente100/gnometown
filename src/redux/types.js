// gnomes
export const fetchRequested = 'FETCH_REQUESTED';
export const fetchRequestedGnomes = 'FETCH_REQUESTED_GNOMES';
export const fetchPendingGnomes = 'FETCH_PENDING_GNOMES';
export const fetchSucceededGnomes = 'FETCH_SUCCEEDED_GNOMES';
export const fetchFailedGnomes = 'FETCH_FAILED_GNOMES';

//filters
export const setAgeFromTo = 'SET_BREWED_FROM_TO';
export const setAgeFrom = 'SET_BREWED_FROM';
export const setAgeTo = 'SET_BREWED_TO';
