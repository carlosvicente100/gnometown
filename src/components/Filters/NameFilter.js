import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

const InputName = styled.input`
  width: -webkit-fill-available;
  height: 40px;
  font-size: 28px;
  margin: 20px;
`

const NameFilter = ({ onChange, current }) => {
  const handleChange = (event) => {
    event.preventDefault()
    onChange(event.target.value)
  }
  return (
    <div>
      <InputName type="text" onChange={handleChange} value={current} placeholder="Gnome Name" />
    </div>
  )
}

NameFilter.propTypes = {}

export default NameFilter
