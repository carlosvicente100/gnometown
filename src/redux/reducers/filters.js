import { fetchSetAgeFromTo, fetchSetJobList, fetchSetName, fetchResetFilters } from '../types'

export const initialState = {
  jobList: [],
  name: '',
  ageTo: {
    years: 0
  },
  ageFrom: {
    years: 0
  }
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchSetJobList:
      console.log(' +-+-+-+-+- action with jobs', action.data.jobs)
      return {
        ...state,
        jobList: action.data.jobs
      }
    case fetchSetName:
      console.log('filter by name', action.data.name)
      return {
        ...state,
        name: action.data.name
      }
    case fetchResetFilters:
      console.log('reset Filters')
      return {
        initialState
      }
    case fetchSetAgeFromTo:
      return {
        ...state,
        ageFrom: action.age.from,
        ageTo: action.age.to
      }
    default:
      return {
        ...state
      }
  }
}
