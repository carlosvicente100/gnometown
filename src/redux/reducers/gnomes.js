import { fetchPendingGnomes, fetchSucceededGnomes, fetchFailedGnomes } from '../types'

export const initialState = {
  pending: false,
  error: null,
  townName: '',
  gnomes: []
}

export default function reducer(state = initialState, action) {
  switch (action.type) {
    case fetchPendingGnomes:
      return {
        ...state,
        pending: true
      }
    case fetchSucceededGnomes:
      const townName = Object.keys(action.data)[0]
      return {
        ...state,
        pending: false,
        townName,
        gnomes: action.data[townName]
      }
    case fetchFailedGnomes:
      return {
        ...state,
        pending: false,
        error: true
      }
    default:
      return {
        ...state
      }
  }
}
