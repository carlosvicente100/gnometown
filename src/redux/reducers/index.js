import { combineReducers } from 'redux';
 import gnomesReducer from './gnomes';
import filtersReducer from './filters';

export default combineReducers({
  gnomesReducer,
  filtersReducer,
});
