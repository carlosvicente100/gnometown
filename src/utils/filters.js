export const getJobsByUsers = (users) => {
  //   console.log('getJobsByUsers', users)
  if (users.length > 0) {
    const profesionList = users
      .map((user) => user.professions)
      .reduce((firstArray, secondArray) => [...firstArray, ...secondArray], [])
    return profesionList.filter((job, index) => profesionList.indexOf(job) === index)
  } else {
    return []
  }
}

export const DEFAULT_VALUE = ''
export const JOBS = 'Jobs'

export const makePagination = (itemsFiltered) => {
  return itemsFiltered.reduce((result, value, index) => {
    if (index % 40 !== 0) {
      result[result.length - 1].push(value)
    } else {
      result.push([value])
    }
    return result
  }, [])
}

export const filterByJob = (jobs, users) => {
  let gnomesWithJobs = users.filter((user) => {
    if (jobs.every((job) => user.professions.includes(job))) {
      return user
    } else {
      return false
    }
  })
  return gnomesWithJobs
}
