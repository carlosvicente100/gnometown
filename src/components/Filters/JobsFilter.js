import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import { JOBS } from '../../utils/filters'

const JobList = styled.div`
font-family:Helvetica;
display: flex;
flex-wrap: wrap;
background-color:#8395a7;
color:white;
text-align:center;
i{
  color:black;
  margin: 5px 0px;
}
}
`
const JobOption = styled.div`
  margin: 5px 15px;
`

const JobsFilter = ({ list, activeList, onSelect, onClear }) => {
  const handleClick = (event) => {
    onSelect(event.target.value, activeList.includes(event.target.value))
  }
  const handleReset = () => {
    onClear(JOBS)
  }

  const OptionList = () => {
    return list.map((option, key) => {
      return (
        <JobOption key={key} data-testid="job-option">
          <input
            type="checkbox"
            id={option}
            value={option}
            checked={activeList.includes(option)}
            onChange={handleClick}
          ></input>
          <label htmlFor={option}> {option}</label>
        </JobOption>
      )
    })
  }

  return (
    <div>
      {list.length > 0 && (
        <JobList>
          <OptionList></OptionList>
          <i data-testid="job-reset" className="fas fa-eraser" onClick={handleReset}></i>
        </JobList>
      )}
    </div>
  )
}

JobsFilter.propTypes = {
  list: PropTypes.array.isRequired,
  activeList: PropTypes.array.isRequired,
  onSelect: PropTypes.func.isRequired,
  onClear: PropTypes.func.isRequired
}

export default JobsFilter
