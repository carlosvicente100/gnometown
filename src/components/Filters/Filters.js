import React, { useState, useEffect } from 'react'
import styled from 'styled-components'
import { useSelector, useDispatch } from 'react-redux'

import { fetchSetJobList, fetchSetName } from '../../redux/types'
import JobsFilter from './JobsFilter'
import NameFilter from './NameFilter'
import { getJobsByUsers, JOBS, ALL } from '../../utils/filters'

//some filters
//by age
//by height
//by weight

const FilterList = styled.div`
  margin: auto;
  text-align: center;
  margin-bottom: 20px;
  display: flex;
  justify-content: center;
  font-size: 28px;
  i {
    margin: 0 20px;
    cursor: pointer;
    border: 1px solid grey;
    border-radius: 10px;
    padding: 5px;
    box-shadow: 1px 1px 5px 0px rgba(50, 50, 50, 0.75);
}

  }
  .open {
    color: #ffbe76;
  }
  .active {
    color: tomato;
  }
`

const Filters = () => {
  const dispatch = useDispatch()
  const { gnomes } = useSelector((state) => state.gnomesReducer)
  const { jobList, name } = useSelector((state) => state.filtersReducer)
  const currentJobList = gnomes.length > 0 ? getJobsByUsers(gnomes) : []
  const [filteredJobs, setFilteredJobs] = useState(jobList)
  const [jobVisible, setJobVisible] = useState(false)
  const [nameFilter, setNameFilter] = useState(name)
  const [nameVisible, setNameVisible] = useState(false)

  const haveJobs = () => jobList.length > 0

  const haveName = () => nameFilter !== ''

  const toggleJobFilter = (flag = !jobVisible) => {
    setJobVisible(flag)
  }
  const toggleNameFilter = (flag = !nameVisible) => {
    setNameVisible(flag)
  }

  const updateJobList = (job, active = false) => {
    if (active) {
      setFilteredJobs(filteredJobs.filter((element) => element !== job))
    } else {
      setFilteredJobs([...filteredJobs, job])
    }
  }

  const filterByName = (name) => {
    setNameFilter(name)
  }

  const removeFilters = () => {
    resetFilter(ALL)
  }

  const resetFilter = (filter) => {
    switch (filter) {
      case JOBS:
        setFilteredJobs([])
        break
      case ALL:
        setFilteredJobs([])
        setNameFilter('') //more filter reset
        break
      default:
        break
    }
  }

  useEffect(() => {
    setJobList()
  }, [filteredJobs])

  useEffect(() => {
    setName()
  }, [nameFilter])

  const setJobList = () => {
    dispatch({
      type: fetchSetJobList,
      data: {
        jobs: filteredJobs
      }
    })
  }
  const setName = () => {
    dispatch({
      type: fetchSetName,
      data: {
        name: nameFilter
      }
    })
  }
  return (
    <>
      <FilterList>
        <i
          data-testid="job-filter-button"
          className={`fas fa-hammer  ${jobVisible ? 'open' : ''} ${haveJobs() ? 'active' : ''} `}
          onClick={() => toggleJobFilter()}
        ></i>
        <i
          data-testid="name-filter-button"
          className={`fas fa-user ${nameVisible ? 'open' : ''} ${haveName() ? 'active' : ''} `}
          onClick={() => toggleNameFilter()}
        ></i>
        {(haveJobs() || haveName()) && (
          <i data-testid="reset-filter-button" className="fas fa-eraser" onClick={removeFilters}></i>
        )}
      </FilterList>
      <div data-testid="filter-container" className="filterActive">
        {jobVisible && (
          <JobsFilter list={currentJobList} activeList={jobList} onSelect={updateJobList} onClear={resetFilter} />
        )}
        {nameVisible && <NameFilter onChange={filterByName} current={nameFilter}></NameFilter>}
      </div>
    </>
  )
}

Filters.propTypes = {}

export default Filters
