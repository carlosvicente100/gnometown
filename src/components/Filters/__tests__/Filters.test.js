import React from 'react'
import { Provider } from 'react-redux'
import { createStore } from 'redux'
import { render as rtlRender, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import 'jest-styled-components'

import { initialState as initialStateGnomes } from '../../../redux/reducers/gnomes'
import { initialState as initialStateFilters } from '../../../redux/reducers/filters'
import reducer from '../../../redux/reducers'

import GnomesDummyList from '../../../fixtures/gnomes'

import Filters from '../Filters'

beforeEach(() => {
  jest.clearAllMocks()
})

const initialReducerState = {
  gnomesReducer: initialStateGnomes,
  filtersReducer: initialStateFilters
}

const profesionListNotEmpty = ['Metalworker', 'Woodcarver']
const nameFilterNotEmpty = 'Tobus'

function render(ui, { initialState, store = createStore(reducer, initialState), ...renderOptions } = {}) {
  function Wrapper({ children }) {
    return <Provider store={store}>{children}</Provider>
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

describe('Filters Component Suite', () => {
  test('Snapshot check', () => {
    expect(render(<Filters />, { initialReducerState }).baseElement).toMatchSnapshot()
  })
})

describe('Filters Structure', () => {
  test('Should have a button for each filter - without active filters', () => {
    const dummyCityName = 'FunkyTown'
    const store = createStore(() => ({
      filtersReducer: {
        jobList: [],
        name: ''
      },
      gnomesReducer: {
        townName: dummyCityName,
        pending: false,
        error: null,
        gnomes: GnomesDummyList[0]
      }
    }))
    const { getByTestId } = render(<Filters />, { store })

    const jobButton = getByTestId('job-filter-button')
    expect(jobButton).toHaveClass('fa-hammer')
    expect(jobButton).not.toHaveClass('open')
    expect(jobButton).not.toHaveClass('active')

    const nameButton = getByTestId('name-filter-button')
    expect(nameButton).toHaveClass('fa-user')
    expect(nameButton).not.toHaveClass('active')
    expect(nameButton).not.toHaveClass('open')

    const filterContainer = getByTestId('filter-container')
    expect(filterContainer.children.length).toBe(0)
  })

  test('Should have Buttons with prop active  - with active filters', async () => {
    const dummyCityName = 'FunkyTown'
    const store = createStore(() => ({
      filtersReducer: {
        jobList: profesionListNotEmpty,
        name: nameFilterNotEmpty
      },
      gnomesReducer: {
        townName: dummyCityName,
        pending: false,
        error: null,
        gnomes: GnomesDummyList[0]
      }
    }))
    const { getByTestId } = render(<Filters />, { store })
    const jobButton = getByTestId('job-filter-button')
    const nameButton = getByTestId('name-filter-button')
    expect(jobButton).not.toHaveClass('open')
    expect(nameButton).not.toHaveClass('open')

    expect(jobButton).toHaveClass('active')
    expect(nameButton).toHaveClass('active')

    fireEvent.click(jobButton)
    fireEvent.click(nameButton)
    expect(jobButton).toHaveClass('open')
    expect(nameButton).toHaveClass('open')
  })
})
