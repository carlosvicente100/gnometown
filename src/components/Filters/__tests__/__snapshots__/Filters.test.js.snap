// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filters Component Suite Snapshot check 1`] = `
.c0 {
  margin: auto;
  text-align: center;
  margin-bottom: 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 28px;
}

.c0 i {
  margin: 0 20px;
  cursor: pointer;
  border: 1px solid grey;
  border-radius: 10px;
  padding: 5px;
  box-shadow: 1px 1px 5px 0px rgba(50,50,50,0.75);
}

.c0 .open {
  color: #ffbe76;
}

.c0 .active {
  color: tomato;
}

<body>
  <div>
    <div
      class="c0"
    >
      <i
        class="fas fa-hammer    "
        data-testid="job-filter-button"
      />
      <i
        class="fas fa-user   "
        data-testid="name-filter-button"
      />
    </div>
    <div
      class="filterActive"
      data-testid="filter-container"
    />
  </div>
</body>
`;
